### PR TITLE
#Mongo_optimization_tasks. AleksandrGogolushko

### PR DESCRIPTION
**Pull request by AleksandrGogolushko**
https://travis-ci.org/github/AleksandrGogolushko/db-assignments/builds/715497166

### CODE REVIEW

File: task/mongo-optimization-task.js [hunk 1]
```diff
@@ -15,7 +15,9 @@ const ObjectId = require('mongodb').ObjectID;
  * Test timeout is increased to 60sec for the function.
  * */
 async function before(db) {
-    await db.collection('opportunities').createIndex({'initiativeId': 1});
+    await db.collection('opportunities').createIndex({ 'initiativeId': 1, "contacts.questions.category_id": 1, "contacts.datePublished": 1 });
+    await db.collection('clientCriteria').createIndex({ "versions.initiativeId": 1 });
+    await db.collection('clientCriteria').createIndex({ "value": 1});
```
Comment (IlyaMokin) [inline task/mongo-optimization-task.js]: in one `$match` you can use only one index a a time. So I don't think that both of the indexes can be used. But you can make compound index (maybe that can be better)

File: task/mongo-optimization-task.js [hunk 2]
```diff
@@ -39,267 +41,265 @@ async function before(db) {
...
-                        "text" : "$criteria.label",
-                        "definition" : {
-                            "$ifNull" : [
-                                "$criteria.versions.definition",
-                                "$criteria.definition"
-                            ]
-                        }
+                {
+                    $match: {
+                        $expr: { $eq: ["$value", "$$v1"] }
```
Comment (IlyaMokin) [inline task/mongo-optimization-task.js]: can you add the query in that lookup in the first `$match` ?

Comment (AleksandrGogolushko) [reply task/mongo-optimization-task.js]: I tried, but it works slower.

Comment (IlyaMokin) [reply task/mongo-optimization-task.js]: 1. That can be ~ the same... but it's weird if it's slower. Also order of fields in index is important. 
`await db.collection('clientCriteria').createIndex({'value': 1,'versions.initiativeId': 1});`
Try vise versa. 

2. limit at the end of pipeline can help in some cases when you need only one item in result of `lookup`

File: task/mongo-optimization-task.js [hunk 3]
```diff
@@ -229,26 +229,23 @@ async function task_3_1(db) {
...
                         "versions.definition": 1,
                     }
-                },
-                {
-                    $match: {
-                        $expr: { $eq: ["$value", "$$v1"] }
-                    }
-                },
+                },{
+                    $limit:1
```
Comment (AleksandrGogolushko) [inline task/mongo-optimization-task.js]: I added a limit and changed a index, but i don't know why it began to work more slowly. On my computer the test passed for 6840ms, in travis - 2552ms. I tried to add limit after first match, after second, at the end of pipeline - similar results.When i deleted limit, but left index, on my computer it was minimum 3050ms, in travis - 1777ms. When i brought back the old index, on my computerit was minimum 2999ms, in travis - 1711ms.
 Maybe am i doing it wrong?

Comment (IlyaMokin) [reply task/mongo-optimization-task.js]: in travis result can be different depending on load of virtual machine. So +-1 sec is ok.
The best approach is execution plan investigating to understand if it work better for specific case.

**Comment by IlyaMokin** [comment]
Thanks for the efforts. The PR is accepted

**Pull request closed without merging**